src/include/elektra.h

```diff
@@ -0,0 +1,367 @@
+ELEKTRA_TAG_DECLARATIONS (kdb_unsigned_long_t, UnsignedLong)
+ELEKTRA_TAG_DECLARATIONS (kdb_long_long_t, LongLong)
+ELEKTRA_TAG_DECLARATIONS (kdb_unsigned_long_long_t, UnsignedLongLong)
+ELEKTRA_TAG_DECLARATIONS (kdb_float_t, Float)
+ELEKTRA_TAG_DECLARATIONS (kdb_double_t, Double)
+
+#if defined(HAVE_SIZEOF_LONG_DOUBLE) && (SIZEOF_LONG_DOUBLE == 16 || SIZEOF_LONG_DOUBLE == 12)
+
+ELEKTRA_TAG_DECLARATIONS (kdb_long_double_t, LongDouble)
+
+#endif // HAVE_SIZEOF_LONG_DOUBLE
+
+ELEKTRA_TAG_DECLARATIONS (int, Enum)
+
+// endregion Tags for built-in types
+
+// region General
+
+Elektra * elektraOpen (const char * application, KeySet * defaults, ElektraError ** error);
+void elektraClose (Elektra * elektra);
+
+size_t elektraArraySize (Elektra * elektra, const char * keyName);
+
+void elektraEnforceTypeMetadata (Elektra * elektra, bool enforceTypeMetadata);
+void elektraFatalErrorHandler (Elektra * elektra, ElektraErrorHandler fatalErrorHandler);
+
+// endregion General
+
+// region Helpers for code generation
+
+Key * elektraFindKey (Elektra * elektra, const char * name, KDBType type);
+Key * elektraFindArrayElementKey (Elektra * elektra, const char * name, size_t index, KDBType type);
+void elektraFatalError (Elektra * elektra, ElektraError * fatalError);
+
+// endregion Helpers for code generation
+
+// region Getters
+
+const char * elektraGetValue (Elektra * elektra, const char * name);
+const char * elektraGetString (Elektra * elektra, const char * keyname);
+kdb_boolean_t elektraGetBoolean (Elektra * elektra, const char * keyname);
+kdb_char_t elektraGetChar (Elektra * elektra, const char * keyname);
+kdb_octet_t elektraGetOctet (Elektra * elektra, const char * keyname);
+kdb_short_t elektraGetShort (Elektra * elektra, const char * keyname);
+kdb_unsigned_short_t elektraGetUnsignedShort (Elektra * elektra, const char * keyname);
+kdb_long_t elektraGetLong (Elektra * elektra, const char * keyname);
+kdb_unsigned_long_t elektraGetUnsignedLong (Elektra * elektra, const char * keyname);
+kdb_long_long_t elektraGetLongLong (Elektra * elektra, const char * keyname);
+kdb_unsigned_long_long_t elektraGetUnsignedLongLong (Elektra * elektra, const char * keyname);
+kdb_float_t elektraGetFloat (Elektra * elektra, const char * keyname);
+kdb_double_t elektraGetDouble (Elektra * elektra, const char * keyname);
+
+#if defined(HAVE_SIZEOF_LONG_DOUBLE) && (SIZEOF_LONG_DOUBLE == 16 || SIZEOF_LONG_DOUBLE == 12)
+
+kdb_long_double_t elektraGetLongDouble (Elektra * elektra, const char * keyname);
+
+#endif
+
+int elektraGetEnumInt (Elektra * elektra, char * keyName);
+
+#define elektraGetEnum(elektra, keyname, enumType) (enumType) elektraGetEnumInt (elektra, keyname)
+
+// endregion Getters
+
+// region Setters
+
+void elektraSetValue (Elektra * elektra, const char * name, const char * value, KDBType type, ElektraError ** error);
+void elektraSetString (Elektra * elektra, const char * keyName, const char * value, ElektraError ** error);
```

Comment:
@markus2330 Wouldn't it be easier to just pass an `ElektraError *` here?

The advantage would be that the error could be allocated by the user:
```c
ElektraError error;
elektraSetString(elektra, "some/key", "some value", &error);
if (elektraErrorCode(error) != 0)
{
    // handle errror
   elektraErrorReset(&error);
}
```

If the user prefers to check the error with `!= NULL`, they could still use `malloc()` to allocate an error and pass that to the function. That way even, if the user does not check for errors, no memory will be leaked.